File: communication_api/REST/cpp/REST.cpp
#include <iostream>
#include <string>
#include <curl/curl.h>
#include <fstream>

class RESTClient {
    public:
        RESTClient(const std::string& base_url) : m_base_url(base_url) {}
        std::string get(const std::string& endpoint);
    private:
        std::string m_base_url;
        static size_t writeCallback(void* contents, size_t size, size_t nmemb, std::string* userp) {
            size_t total_size = size * nmemb;
            userp->append((char*)contents, total_size);
            return total_size;
        }
};

std::string RESTClient::get(const std::string& endpoint) {
    CURL* l_curl = curl_easy_init();
    std::string full_url = m_base_url + endpoint;
    std::string response_data;
    if (l_curl) {
        CURLcode result;
        curl_easy_setopt(l_curl, CURLOPT_URL, full_url.c_str());
        curl_easy_setopt(l_curl, CURLOPT_WRITEFUNCTION, writeCallback);
        curl_easy_setopt(l_curl, CURLOPT_WRITEDATA, &response_data);
        result = curl_easy_perform(l_curl);
        if (result != CURLE_OK) {
            std::cerr << "CURL error: " << curl_easy_strerror(result) << std::endl;
        }
        curl_easy_cleanup(l_curl);
    }
    return response_data;
}

int main()
{
    std::string API_KEY="";
    FILE* file = fopen("../key.txt", "r");
    if (file) {
        char buffer[256];
        while (fgets(buffer, sizeof(buffer), file)) {
            API_KEY.append(buffer);
        }
        fclose(file);
    } else {
        std::cerr << "Failed to open file." << std::endl;
    }
    std::string CITY_NAME;
    std::cout << "Enter City Name: ";
    std::cin >> CITY_NAME;

    curl_global_init(CURL_GLOBAL_DEFAULT);
    RESTClient client("https://api.openweathermap.org/data/2.5/weather?q=" + CITY_NAME + "&appid=" + API_KEY + "&units=imperial");
    std::string response = client.get("/data");
    // [TODO] Parse the response JSON and extract required information
    std::cout << "Response: " << response << std::endl;
    curl_global_cleanup();
    return 0;
}

File: communication_api/REST/python/REST.py
import requests

class RESTClient:
    def __init__(self, base_url):
        self.base_url = base_url

    def get(self, endpoint, params=None, headers=None):
        url = f"{self.base_url}/{endpoint}"
        response = requests.get(url, params=params, headers=headers)
        return self._handle_response(response)

    def post(self, endpoint, data=None, json=None, headers=None):
        url = f"{self.base_url}/{endpoint}"
        response = requests.post(url, data=data, json=json, headers=headers)
        return self._handle_response(response)

    def put(self, endpoint, data=None, json=None, headers=None):
        url = f"{self.base_url}/{endpoint}"
        response = requests.put(url, data=data, json=json, headers=headers)
        return self._handle_response(response)

    def delete(self, endpoint, headers=None):
        url = f"{self.base_url}/{endpoint}"
        response = requests.delete(url, headers=headers)
        return self._handle_response(response)

    def _handle_response(self, response):
        if response.status_code >= 200 and response.status_code < 300:
            try:
                return response.json()
            except ValueError:
                return response.text
        else:
            response.raise_for_status()

def main():
    MY_API_KEY = open("../key.txt", "r").read().strip()
    CITY_NAME = input("Enter City Name: ")
    client = RESTClient(f"https://api.openweathermap.org/data/2.5/weather?q={CITY_NAME}&appid={MY_API_KEY}&units=imperial")
    response = client.get("")
    try:
        print(f"Temperature in {CITY_NAME}: {response['main']['temp']}°F")
    except KeyError:
        print(f"Error: {response}")

if __name__ == "__main__":
    main()